File: apps/ios/OpenFrame/Repository/KioskRepository.swift
import Foundation

final class KioskRepository: Sendable {
    private let apiClient: APIClient

    init(apiClient: APIClient) {
        self.apiClient = apiClient
    }

    func getKiosks() async -> Result<[Kiosk], Error> {
        do {
            let dtos: [KioskDTO] = try await apiClient.requestData(.get, path: "/api/v1/kiosks")
            return .success(dtos.map { $0.toDomain() })
        } catch {
            return .failure(error)
        }
    }

    func getKiosk(id: String) async -> Result<Kiosk, Error> {
        do {
            let dto: KioskDTO = try await apiClient.requestData(.get, path: "/api/v1/kiosks/\(id)")
            return .success(dto.toDomain())
        } catch {
            return .failure(error)
        }
    }

    func sendCommand(kioskId: String, type: String, payload: [String: String]? = nil) async -> Result<Void, Error> {
        do {
            let body = KioskCommandRequest(type: type, payload: payload)
            try await apiClient.requestVoid(.post, path: "/api/v1/kiosks/\(kioskId)/command", body: body)
            return .success(())
        } catch {
            return .failure(error)
        }
    }

    func refreshKiosk(id: String) async -> Result<Void, Error> {
        do {
            try await apiClient.requestVoid(.post, path: "/api/v1/kiosks/\(id)/refresh")
            return .success(())
        } catch {
            return .failure(error)
        }
    }

    func getSavedFiles(kioskId: String) async -> Result<[KioskSavedFile], Error> {
        do {
            let dtos: [KioskSavedFileDTO] = try await apiClient.requestData(.get, path: "/api/v1/kiosks/\(kioskId)/files")
            return .success(dtos.map { $0.toDomain() })
        } catch {
            return .failure(error)
        }
    }

    func deleteSavedFile(kioskId: String, fileId: String) async -> Result<Void, Error> {
        do {
            try await apiClient.requestVoid(.delete, path: "/api/v1/kiosks/\(kioskId)/files/\(fileId)")
            return .success(())
        } catch {
            return .failure(error)
        }
    }
}

File: apps/android/app/src/main/java/us/openframe/app/data/repository/KioskRepository.kt
package us.openframe.app.data.repository

import us.openframe.app.data.remote.api.KioskApi
import us.openframe.app.data.remote.dto.KioskCommandRequest
import us.openframe.app.data.remote.dto.KioskDashboardDto
import us.openframe.app.data.remote.dto.KioskDto
import us.openframe.app.domain.model.Kiosk
import us.openframe.app.domain.model.KioskDashboard
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class KioskRepository @Inject constructor(
    private val kioskApi: KioskApi,
) {
    suspend fun getKiosks(): Result<List<Kiosk>> {
        return try {
            val response = kioskApi.getKiosks()
            if (response.isSuccessful) {
                val kiosks = (response.body()?.data ?: emptyList()).map { it.toDomain() }
                Result.success(kiosks)
            } else {
                Result.failure(Exception("Failed to fetch kiosks"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun getKiosk(id: String): Result<Kiosk> {
        return try {
            val response = kioskApi.getKiosk(id)
            if (response.isSuccessful) {
                val kiosk = response.body()?.data?.toDomain() ?: return Result.failure(Exception("No data"))
                Result.success(kiosk)
            } else {
                Result.failure(Exception("Failed to fetch kiosk"))
            }
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun sendCommand(kioskId: String, type: String, payload: Map<String, Any>? = null): Result<Unit> {
        return try {
            val response = kioskApi.sendCommand(kioskId, KioskCommandRequest(type, payload))
            if (response.isSuccessful) Result.success(Unit)
            else Result.failure(Exception("Command failed"))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    suspend fun refreshKiosk(kioskId: String): Result<Unit> {
        return try {
            val response = kioskApi.refreshKiosk(kioskId)
            if (response.isSuccessful) Result.success(Unit)
            else Result.failure(Exception("Refresh failed"))
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}

private fun KioskDto.toDomain() = Kiosk(
    id = id, name = name, isActive = isActive, displayMode = displayMode,
    displayType = displayType, colorScheme = colorScheme,
    screensaverEnabled = screensaverEnabled,
    dashboards = dashboards?.map { it.toDomain() } ?: emptyList(),
    lastAccessedAt = lastAccessedAt,
)

private fun KioskDashboardDto.toDomain() = KioskDashboard(
    id = id, type = type, name = name, icon = icon, pinned = pinned,
)
